### DISCUSSION TITLE
[SNAT] How to change source ip address from vpn to a actual internal ip ?

### DISCUSSION THREAD

**Discussion by Hacksign** [strongswan/strongswan#2111] [Q&A] [answered]
Hi there,

## Background

I've managed to setup a IPsec VPN network these days.
I'm not familiar with network stuffs, but I guess it is a SNAT problem after I did some sniffering.

## Below is some basic information:

`SiteA` in `LAN1`:
```
internal ip: 10.1.2.3/24
public ip: 1.2.3.4
ipsec vpn ip: 192.168.100.17/28
```

`SiteB` in `LAN1`:
```
internal ip: 10.1.2.7/24
no public ip.
```

`SiteC` in `LAN2`:
```
internal ip: 192.168.1.121/24
public ip: 5.6.7.8
ipsec vpn ip: 192.168.100.1/28
```

Now there is a http service hosted on `10.1.2.7/24`, which is `SiteB`, I want access `http://SiteB` on `SiteC`.

The traffic flow should like this :
```
     SiteC ----> Internet(IPSec Tunnel) ----> SiteA ----> SiteB
|<---LAN2--->|                               |<-----LAN1----->|
```

I've sniffered the packets with tcpdump on `SiteA`, and found that the request packet (TCP SYN, the first packet) can be reached from `SiteC` to `SiteA`, the problem is source address is `192.168.100.0` (first address of `SiteC`'s VPN range):

```
[root@SiteA user1]# tcpdump -n -i end0 port 8000
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on end0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
19:15:40.730945 IP 192.168.100.0.61820 > 10.1.2.7.8000: Flags [S], seq 2042563908, win 64240, options [mss 1360,nop,wscale 8,nop,nop,sackOK], length 0
19:15:40.762088 IP 192.168.100.0.61821 > 10.1.2.7.8000: Flags [S], seq 2213759697, win 64240, options [mss 1360,nop,wscale 8,nop,nop,sackOK], length 0
19:15:40.993900 IP 192.168.100.0.61822 > 10.1.2.7.8000: Flags [S], seq 2887494785, win 64240, options [mss 1360,nop,wscale 8,nop,nop,sackOK], length 0
19:15:44.943436 IP 192.168.100.0.61863 > 10.1.2.7.8000: Flags [S], seq 1725618234, win 64240, options [mss 1360,nop,wscale 8,nop,nop,sackOK], length 0
19:15:44.965016 IP 192.168.100.0.61864 > 10.1.2.7.8000: Flags [S], seq 4158635734, win 64240, options [mss 1360,nop,wscale 8,nop,nop,sackOK], length 0
```

I dont see the request packets to 8000 by command `tcpdump -n -i eth0 port 8000` on SiteB.(`eth0` is `10.1.2.7/24`)

`SiteC` is configured by `/etc/ipsec.conf` and `SiteA` is configured by '/etc/swanctl/swanctl.conf', these 2 hosts can communicate normally.

Here is the configuraion:

## Configurations

ipsec.conf
```
conn SiteC
  keyexchange=ikev2
  left=%any
  leftsubnet=192.168.100.0/24,192.168.1.0/24
  leftcert=vpn-ca.crt
  rightauth=eap-mschapv2
  rightsendcert=never
  eap_identity=%any
  right=%any
  rightid=vpn-user1
  rightsubnet=192.168.100.0/28,10.0.0.0/8
  rightsourceip=192.168.100.1-192.168.100.15
  auto=route
```

swanctl.conf
```
connections {
   site-c-vpn {
       encap = yes
       mobike = yes
       version = 2
       keyingtries = 0
       dpd_delay = 30
       dpd_timeout = 90
       vips = 192.168.100.1
       remote_addrs = 5.6.7.8

       local {
           auth = eap-mschapv2
           id = vpn-user1
           eap_id = vpn-user1
       }

       remote {
           certs = vpn-ca.pem 
           id = example.com
       }

       children {
           site-c-vpn-children {
               local_ts = 192.168.100.0/24,10.0.0.0/8
               remote_ts = 192.168.100.0/24,192.168.1.0/24
               start_action = start
               dpd_action = restart
               close_action = restart
           }
       }
   }
}

```

## Questions:

1. How to rewrite requests' source ip from `192.168.100.0` to the proper ip (`10.1.2.3` in my case) ?
2. My original goal is access internal resources in `LAN1` with `SiteC`, if `Question 1` can be solved, can I achieve my goal ?

**Comment by tobiasbrunner** [comment]
I suggest you read [the docs](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html).

**Reply by Hacksign** [reply]
I've read this doc, but still no clue how to solve my problem.
Can you give me some examples ?

**Comment by Hacksign** [comment, marked as answer]
OK figured out by myself ...

Add a SNAT iptable rule solved my problem:

```
iptables -t nat -I POSTROUTING -s 192.168.100.0/24 -j MASQUERADE
```

Where `192.168.100.0/24` is the VPN virtual ip address range.

Dont forget open ip_forwart for ipv4 (ipv6 should be also configured if you use ipv6 address) or the iptable rule will not work:

```
echo 1 > /proc/sys/net/ipv4/ip_forward
```
